core/cppgc.rs

```diff
@@ -69,15 +215,25 @@ pub fn wrap_object<'a, T: GarbageCollected + 'static>(
   let member = unsafe {
     v8::cppgc::make_garbage_collected(
       heap,
-      CppGcObject {
-        tag: TypeId::of::<T>(),
-        member: t,
-      },
+      PrototypeChainStore([
```

Comment:
do normal objects need this? its unfortunate to have double the allocations. could we instead have two tags, one for plain objects and one for prototype backed objects? plan objects can continue to use wrap/unwrap, and prototype backed objects can use wrap2/unwrap2.

Reply:
yeah I think we could do something ala:

```rust
let ptr = if T::prototype_index().is_none() {
  v8::Object::unwrap::<CPPGC_TAG, CppGcObject<T>>(isolate, obj)
} else {
  let proto_chain = v8::Object::unwrap::<CPPGC_TAG, PrototypeChainStore>(isolate, obj)?;

  let obj = proto_chain.0[proto_index]
    .as_ref()?
    .downcast::<T>()?;
};
```

Reply:
@littledivy ^^

Reply:
That could work with a custom DST. A safe enum will require heap indirection or just have the same size.

I'll give the DSTs a shot, that might also improve the `3` protochain limitation we have since we could just size it at runtime.

```
size: u8
data: [CppGcObject]
```

Reply:
if you can solve it with one system that's cool, but i just meant have two different wrap mechanisms. the default one would be the existing one, and you can opt into a prototype mode if you're using that.

Reply:
Ok, I made the prototype chain opt-in with `#[op2(base)]` on the base class otherwise it is treated as a basic object with a single CppGcObject allocation.

Reply:
this is pretty good, but not exactly what i meant: i would split into two different things: wrap and wrap with prototype. specifically leave the existing wrap/unwrap functions untouched, and don't require structs using them to implement the prototype trait. and then you can have the new wrapper functions for structs using prototypes.

Reply:
self would always be with proto. plain fn would be without proto unless you do like #[cppgc(proto)] or something

Reply:
> self would always be with proto.

don't we want it to without proto for single objects tho?

```rust
// want this to be without proto:
#[op2]
impl A {}

// want this to be with proto:
#[op2(base)]
impl Base {}
#[op2(inherit = Base)]
impl B {}
```

So maybe we do this: by default `self` and `#[cppgc]` are no-proto. And we do proto only when `base` and `inherit` attribute or`#[cppgc(proto)]` is used.

Downside is that devs will have to keep track of what has proto and what doesn't - which was already done for then with the default `PrototypeChain` trait.

Reply:
ultimately I'm just aiming to keep the complexity down for cppgc backed resources. prototype stuff has a lot of its own codegen already so it should be free to use new apis I think.

Reply:
PTAL https://github.com/denoland/deno_core/pull/1047/commits/4dc27fb85f4a19d5f10d10b38c6ee3fd707a65cc

didn't implement `#[cppgc(proto)]` cuz we don't need it right now.